### PR TITLE
Migrate S3854 to eslint implementation

### LINKED ISSUE #2279: Migrate rule S3854
https://eslint.org/docs/rules/constructor-super, https://eslint.org/docs/rules/no-this-before-super, there are two rules in eslint

### PR DESCRIPTION
**Pull request by guillaume-dequenne-sonarsource**
Fixes #2279 
Two possibles approaches are proposed:
1. In `linter.ts`, use existing linter to retrieve `constructor-super` and `no-this-before-super` eslint rules and merge them through a wrapper function.
2. Create `super-invocation` as its own rule and use a local linter to retrieve the 2 eslint rules and merge them.

There are a few FN/FP compared to the previous Java implementation but it seems acceptable overall.

### CODE REVIEW

**Review by saberduck — COMMENTED** [review]
I think we could try to implement our rule which would delegate to the rules from eslint

**Review by saberduck — APPROVED** [review]

File: its/sources/file-for-rules/S3854.js [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+    super.x = 1;
+  }
+}
+
+class A2 {
+  constructor() {
+    super();                 // Noncompliant {{super() can only be invoked in a derived class constructor.}}
+//  ^^^^^
+    this.f = function() {
+      super();               // Noncompliant {{super() can only be invoked in a derived class constructor.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:26]: I think we could change comment from 'Noncompliant' to 'FN'

File: its/sources/file-for-rules/S3854.js [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+    }
+    this.g = (function() {
+      super();               // Noncompliant {{super() can only be invoked in a derived class constructor.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:29]: same here

File: its/sources/file-for-rules/S3854.js [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+    })(1)
+  }
+  bar() {
+    super();                 // Noncompliant {{super() can only be invoked in a derived class constructor.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:33]: same here

File: its/sources/file-for-rules/S3854.js [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+  }
+}
+
+class B2a extends A2 {
+  constructor() {
+    super();                 // OK
+    this.f = function() {
+      super();               // Noncompliant {{super() can only be invoked in a derived class constructor.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:41]: same here

File: its/sources/file-for-rules/S3854.js [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+    }
+  }
+}
+
+var B2b = class extends A2 {
+  constructor() {
+    super();                 // OK
+    this.f = function() {
+      super();               // Noncompliant {{super() can only be invoked in a derived class constructor.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:50]: same here

File: its/sources/file-for-rules/S3854.js [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+    }
+  }
+}
+
+class A3 extends Unknown {
+  constructor() {
+    super(x);                // OK
+  }
+  bar() {
+    super(x);                // Noncompliant {{super() can only be invoked in a derived class constructor.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:60]: same here

File: its/sources/file-for-rules/S3854.js [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+
+class A4 extends Unknown {
+  constructor() {
+    super(x);                // OK
+  }
+}
+
+class A5 extends null {
+  constructor() {
+    super();                 // NOK, but no solution: super() -> TypeError, whereas no super() -> ReferenceError
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:72]: I think comment should be updated here

File: its/sources/file-for-rules/S3854.js [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+  }
+}
+
+class A6 extends null {
+  constructor() {
+    // no super()            // NOK, but no solution: super() -> TypeError, whereas no super() -> ReferenceError
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:78]: I think comment should be updated here

File: its/sources/file-for-rules/S3854.js [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+  }
+}
+
+function foo() {
+  super();                   // Noncompliant {{super() can only be invoked in a derived class constructor.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:83]: Noncompliant -> FN

File: its/sources/file-for-rules/S3854.js [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+}
+
+function constructor() {
+  super();                   // Noncompliant {{super() can only be invoked in a derived class constructor.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:87]: Noncompliant -> FN

File: its/sources/file-for-rules/S3854.js [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+}
+
+super();                     // Noncompliant {{super() can only be invoked in a derived class constructor.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:90]: Noncompliant -> FN

File: its/sources/file-for-rules/S3854.js [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+//S ^^^^                     ID1
+    super.x = x + y;
+//S ^^^^^                    ID1
+    this.foo();
+//S ^^^^                     ID1
+    bar(this);
+//S     ^^^^                 ID1
+    bar(super.x);
+//S     ^^^^^                ID1
+    super(x);                // Noncompliant [[ID=ID1]] {{super() must be invoked before "this" or "super" can be used.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:143]: Reporting is different, eslint raises at line 133, 135, 137, 139, 141

File: its/sources/file-for-rules/S3854.js [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+//  ^^^^^
+    super.foo();
+  }
+}
+
+class B20c extends A20 {
+  constructor(x, y) {
+    this.foo(
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:151]: Primary location is here

File: its/sources/file-for-rules/S3854.js [hunk 14]
```diff
@@ -0,0 +1,406 @@
...
+
+class C30 extends B30 {
+  constructor(x, y) {
+    super(x);                // OK
+  }
+}
+
+class B30b extends A30 {
+  constructor(x) {
+    super(x);                // Noncompliant {{super() must be invoked with 0 arguments.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:191]: Noncompliant -> FN

File: its/sources/file-for-rules/S3854.js [hunk 15]
```diff
@@ -0,0 +1,406 @@
...
+  }
+}
+
+class B30c extends A30 {
+  constructor(x) {
+    super(x, y);             // Noncompliant {{super() must be invoked with 0 arguments.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:197]: Noncompliant -> FN

File: its/sources/file-for-rules/S3854.js [hunk 16]
```diff
@@ -0,0 +1,406 @@
...
+}
+
+class A32 {
+    constructor(x, y, z) {}
+//S            ^^^^^^^^^ 32
+}
+
+class B32 extends A32 {
+  constructor() {
+    super(x, y);             // Noncompliant [[ID=32]] {{super() must be invoked with 3 arguments.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:208]: Noncompliant -> FN

File: its/sources/file-for-rules/S3854.js [hunk 17]
```diff
@@ -0,0 +1,406 @@
...
+
+  class B35a extends A35() {
+    constructor(x) {
+      super(x);              // OK
+    }
+  }
+
+  class B35b extends A35 {
+    constructor(x,y) {
+      super(x, y);           // Noncompliant {{super() must be invoked with 1 argument.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:230]: Noncompliant -> FN

File: its/sources/file-for-rules/S3854.js [hunk 18]
```diff
@@ -0,0 +1,406 @@
...
+    }
+  }
+
+  class D34a extends A33 {
+    constructor(x) {
+      super(x);              // Noncompliant {{super() must be invoked with 3 arguments.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:236]: Noncompliant -> FN

File: its/sources/file-for-rules/S3854.js [hunk 19]
```diff
@@ -0,0 +1,406 @@
...
+    }
+  }
+
+  class D34b extends B32 {
+    constructor(x) {
+      super(x, y);           // Noncompliant {{super() must be invoked with 0 arguments.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:242]: Noncompliant -> FN

File: its/sources/file-for-rules/S3854.js [hunk 20]
```diff
@@ -0,0 +1,406 @@
...
+    var array = [{
+      constructor() {   // OK
+      }
+    }];
+  }
+}
+
+var top_level_object = {
+  constructor() {         // OK
+    super();              // Noncompliant
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:345]: Noncompliant -> FN

File: its/sources/file-for-rules/S3854.js [hunk 21]
```diff
@@ -0,0 +1,406 @@
...
+
+class Derived4 extends Super {
+    constructor() {
+        super(1, 2, 3, 4);
+    }
+}
+
+class Derived5 extends Super {
+    constructor() {
+        super(1, 2, 3, 4, 5); // Noncompliant
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:398]: Noncompliant -> FN

File: javascript-checks/src/test/java/org/sonar/javascript/checks/SuperInvocationCheckTest.java [hunk 22]
```diff
@@ -1,36 +0,0 @@
...
-import org.sonar.javascript.checks.verifier.JavaScriptCheckVerifier;
-import org.sonar.plugins.javascript.api.JavaScriptCheck;
-
-public class SuperInvocationCheckTest {
-
-  private final JavaScriptCheck check = new SuperInvocationCheck();
-
-  @Test
-  public void test() {
-    JavaScriptCheckVerifier.verify(check, new File("src/test/resources/checks/superInvocation.js"));
```
Comment (andrea-guarino-sonarsource) [inline javascript-checks/src/test/java/org/sonar/javascript/checks/SuperInvocationCheckTest.java:33]: I think you forgot to delete `src/test/resources/checks/superInvocation.js`

File: its/sources/file-for-rules/S3854.js [hunk 23]
```diff
@@ -0,0 +1,406 @@
...
+var B1b = class extends A1 {
+  constructor() {
+    super();                 // OK
+    super.x = 1;
+  }
+}
+
+class A2 {
+  constructor() {
+    super();                 // Noncompliant {{super() can only be invoked in a derived class constructor.}}
```
Comment (andrea-guarino-sonarsource) [inline its/sources/file-for-rules/S3854.js:23]: This issue is reported as a parsing error by [eslint](https://eslint.org/demo#eyJ0ZXh0IjoiY2xhc3MgQSB7XG4gY29uc3RydWN0b3IoKSB7XG4gICBzdXBlcigpO1xuIH1cbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjJ9LCJlbnYiOnt9fX0=)
I'm not sure if we will report twice when we'll migrate "ParsingErrorCheck"

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java [hunk 24]
```diff
@@ -19,323 +19,18 @@
...
-import org.sonar.plugins.javascript.api.tree.expression.SuperTree;
-import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
-import org.sonar.plugins.javascript.api.visitors.DoubleDispatchVisitor;
-import org.sonar.plugins.javascript.api.visitors.DoubleDispatchVisitorCheck;
-import org.sonar.plugins.javascript.api.visitors.IssueLocation;
-import org.sonar.plugins.javascript.api.visitors.PreciseIssue;
+import org.sonar.plugins.javascript.api.TypeScriptRule;
 
 @JavaScriptRule
+@TypeScriptRule
```
Comment (andrea-guarino-sonarsource) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java:28]: Does it make sense to activate for TS? I think it's already reported by the compiler

**Review by andrea-guarino-sonarsource — COMMENTED** [review]
Overall LGTM, I think we should update comments in IT to document the introduced FN and changes in reporting.
Also, I'm not sure we should active this rule for TypeScript

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java (hunk 24 above)
Comment (guillaume-dequenne-sonarsource) [reply javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java:28]: Indeed, I'll just keep it for JavaScript.

File: its/sources/file-for-rules/S3854.js (hunk 23 above)
Comment (guillaume-dequenne-sonarsource) [reply its/sources/file-for-rules/S3854.js:23]: My understanding is that if there is a failing error, we fall back to another parser, which might raise this issue (and not the parsing error). If a parsing error remains, it should be the only thing reported on the file, isn't that so?

Comment (andrea-guarino-sonarsource) [reply its/sources/file-for-rules/S3854.js:23]: Indeed, I think you're right. If there is a parsing error, it should be the only issue in the file.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java (hunk 24 above)
Comment (guillaume-dequenne-sonarsource) [reply javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java:28]: I'll actually re-enable the rule for TypeScript as it was already enabled prior to this PR.

**Pull request merged**